File: shared/src/main/java/com/openmeet/shared/data/message/MessageDAO.java
package com.openmeet.shared.data.message;

import com.openmeet.shared.data.storage.DAO;
import com.openmeet.shared.data.storage.GenericDAO;
import com.openmeet.shared.data.storage.SQLDAO;
import com.openmeet.shared.exceptions.InvalidPrimaryKeyException;

import java.sql.SQLException;
import java.util.List;
import java.util.HashMap;
import javax.sql.DataSource;

public class MessageDAO extends SQLDAO implements DAO<Message> {

    public MessageDAO(DataSource source) {
        super(source);
    }

    @Override
    public List<Message> doRetrieveByCondition(String condition) throws SQLException {
        return GenericDAO.genericDoRetrieveByCondition(Message.MESSAGE, condition, new MessageExtractor(), source);
    }

    @Override
    public List<Message> doRetrieveByCondition(String condition, int row_count) throws SQLException {
        return GenericDAO.genericDoRetrieveByCondition(Message.MESSAGE, condition + " LIMIT " + row_count,
                new MessageExtractor(), source);
    }

    @Override
    public List<Message> doRetrieveByCondition(String condition, int offset, int row_count) throws SQLException {
        return GenericDAO.genericDoRetrieveByCondition(Message.MESSAGE, condition + " LIMIT " + offset + ", " + row_count,
                new MessageExtractor(), source);
    }

    @Override
    public Message doRetrieveByKey(String key) throws SQLException, InvalidPrimaryKeyException {

        if (key == null) {
            throw new InvalidPrimaryKeyException(key);
        }
        List<Message> message = doRetrieveByCondition(
                String.format("%s = '%s'", Message.MESSAGE_ID, key)
        );
        return message.isEmpty() ? null : message.get(0);
    }

    @Override
    public List<Message> doRetrieveAll() throws SQLException {
        return doRetrieveByCondition("TRUE");
    }

    @Override
    public List<Message> doRetrieveAll(int row_count) throws SQLException {
        return doRetrieveByCondition("TRUE LIMIT " + row_count);
    }

    @Override
    public List<Message> doRetrieveAll(int offset, int row_count) throws SQLException {
        return doRetrieveByCondition("TRUE LIMIT " + offset + ", " + row_count);
    }

    @Override
    public boolean doSave(Message obj) throws SQLException {
        return GenericDAO.genericDoSave(Message.MESSAGE, obj.toHashMap(), this.source);
    }

    @Override
    public boolean doSave(HashMap<String, ?> values) throws SQLException {
        return GenericDAO.genericDoSave(Message.MESSAGE, values, this.source);
    }

    @Override
    public boolean doUpdate(HashMap<String, ?> values, String condition) throws SQLException {
        return GenericDAO.genericDoUpdate(Message.MESSAGE, condition, values, this.source);
    }

    @Override
    public boolean doSaveOrUpdate(Message obj) throws SQLException {

        if (doRetrieveByKey(String.valueOf(obj.getId())) == null)
            return doSave(obj);

        return doUpdate(obj.toHashMap(), String.format("%s = '%s'", Message.MESSAGE_ID, obj.getId()));
    }

    @Override
    public boolean doDelete(String condition) throws SQLException {
        return GenericDAO.genericDoDelete(Message.MESSAGE, condition, source);
    }
}

File: mobileapp/app/src/main/java/com/openmeet/data/message/MessageDAO.kt
package com.openmeet.data.message

import com.openmeet.shared.data.storage.DAO
import com.openmeet.shared.data.storage.SQLDAO
import java.util.HashMap
import javax.sql.DataSource

class MessageDAO(source: DataSource) : SQLDAO(source), DAO<Message> {

    override fun doRetrieveByCondition(condition: String?): MutableList<Message> {
        return genericDoRetrieveByCondition(
            Message.MESSAGE,
            condition,
            MessageExtractor(),
            source
        )
    }

    override fun doRetrieveByKey(key: String?): Message? {

        if (key == null) {
            throw IllegalArgumentException("null")
        }

        val message = doRetrieveByCondition("${Message.MESSAGE}.id = '$key'")

        return if (message.isEmpty()) null else message[0]
    }

    override fun doRetrieveAll(): MutableList<Message> {
        return doRetrieveByCondition("TRUE")
    }

    override fun doRetrieveAll(row_count: Int): MutableList<Message> {
        return doRetrieveByCondition("TRUE LIMIT $row_count")
    }

    override fun doRetrieveAll(offset: Int, row_count: Int): MutableList<Message> {
        return doRetrieveByCondition("TRUE LIMIT $offset, $row_count");
    }

    override fun doSave(obj: Message?): Boolean {
        return genericDoSave(Message.MESSAGE, obj?.toHashMap(), source)
    }

    override fun doUpdate(values: HashMap<String, *>?, condition: String?): Boolean {
        return genericDoUpdate(Message.MESSAGE, condition, values, source)
    }

    override fun doSaveOrUpdate(obj: Message?): Boolean {

        if (doRetrieveByKey(obj?.id.toString()) == null) {
            return doSave(obj)
        }
        return doUpdate(obj?.toHashMap(), "${Message.MESSAGE}.id = '${obj?.id}'")
    }

    override fun doDelete(condition: String?): Boolean {
        return genericDoDelete(Message.MESSAGE, condition, source)
    }
}